File: ghidra_scripts/ghidra_extract.java
/*
Copyright 2023 National Technology & Engineering Solutions
of Sandia, LLC (NTESS). Under the terms of Contract DE-NA0003525 with NTESS,
the U.S. Government retains certain rights in this software.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

import java.util.Iterator;

import ghidra.app.script.GhidraScript;
import ghidra.program.model.mem.*;
import ghidra.program.model.lang.*;
import ghidra.program.model.pcode.*;
import ghidra.program.model.util.*;
import ghidra.program.model.reloc.*;
import ghidra.program.model.data.*;
import ghidra.program.model.block.*;
import ghidra.program.model.symbol.*;
import ghidra.program.model.scalar.*;
import ghidra.program.model.listing.*;
import ghidra.program.model.address.*;

import ghidra.app.decompiler.*;

public class ghidra_extract extends GhidraScript {

    Address baseAddress;
    public String fixAddress(Address a) {
        return "" + a.subtract(baseAddress);
    }

    public void run() throws Exception {
        SimpleBlockModel sbm = new SimpleBlockModel(currentProgram);

        // Set up decompiler interface for our program
        DecompInterface ifc = new DecompInterface();
        ifc.openProgram(currentProgram);

        baseAddress = currentProgram.getImageBase();
        String format = currentProgram.getExecutableFormat();
        println("\nFile format: " + format + "\nImage base: " + currentProgram.getImageBase());

        // Process Instructions and dump with RESULTI
        InstructionIterator ii = currentProgram.getListing().getInstructions(true);
        print ( "\nRESULTM: [\"" + currentProgram.getImageBase() + "\"]\n\n");
        while (ii.hasNext() ) {
            Instruction in = ii.next();
            String output = "[";
            output += "\"" + in.getAddress() + "\", \"";
            String mnemonic = in.getMnemonicString();
            for (int i = 0; i < in.getNumOperands(); i++) {
                if (i == 0) {
                    mnemonic += "  ";
                } else {
                    mnemonic += ",";
                }

                mnemonic += in.getDefaultOperandRepresentation(i);
            }
            output += mnemonic + "\", ";

            // Add basic block parent to instruction
            output += "\"" + fixAddress(sbm.getFirstCodeBlockContaining(in.getAddress(), monitor).getFirstStartAddress()) + "\"]";
            print ( "\nRESULTI: " + output + "\n\n");
        }



       
    }
}

File: ghidra_scripts/ghidra_extract.py
from __future__ import print_function

# `currentProgram` or `getScriptArgs` function is contained in `__main__`
# actually you don't need to import by yourself, but it makes much "explicit"
import __main__ as ghidra_app


import json , collections
from collections import OrderedDict

from ghidra.program.model.data import Pointer, StructureDataType
import re

import hashlib
# from ghidra.program.model.listing import OperandType
from ghidra.program.model.lang import OperandType




def run():


    output_path = str(getScriptArgs()[0])
    # start_file_offset = 0  
    # ghidra_app.currentProgram.setImageBase(ghidra_app.getCurrentProgram().getAddressFactory().getDefaultAddressSpace().getAddress(start_file_offset), True)
    baseAddress = ghidra_app.currentProgram.getImageBase()


    all_instructions = ghidra_app.currentProgram.getListing().getInstructions(True)
    # self.print(f"\nRESULTM: [\"{self.base_address}\"]\n\n")
    addr_inst_dict =  OrderedDict()
    for instruction in all_instructions:

        address = int(instruction.getAddress().subtract(0).toString(),16) #int( inst.getAddress().subtract(0).toString(),16)

        mnemonic = instruction.getMnemonicString()
        operands = []
        for i in range(instruction.getNumOperands()):
            operands.append(instruction.getDefaultOperandRepresentation(i))

        mnemonic_with_operands = mnemonic + ' ' + ', '.join(operands)

        # print(address , mnemonic_with_operands)
        inst_file_offset = ghidra_app.currentProgram.getMemory().getAddressSourceInfo(instruction.getAddress()).getFileOffset()
        # if int(inst_file_offset)!=int(address):
        #     print(int(inst_file_offset) , int(address))
        #     print('Offset error ',  ghidra_app.currentProgram.getImageBase())
        #     # return
        # else:
        #     print(inst_file_offset ,address)
        addr_inst_dict[int(inst_file_offset)] = mnemonic_with_operands


    # sorted_keys = sorted(addr_inst_dict.keys())
    # # Reconstruct the dictionary with sorted keys
    # sorted_dict = {key: addr_inst_dict[key] for key in sorted_keys}

    with open(output_path, 'w') as fw:
        json.dump(addr_inst_dict, fw, indent=2)

# To use this script in Ghidra, save it as a .py file and run it within Ghidra's script manager.



# Starts execution here
if __name__ == '__main__':
    run()
    